File: cxx/rest/RESTHandler.cpp
/**
 * @file RESTHandler.cpp
 * @author Ronald T. Fernandez
 * @version 1.0
 */

#include <cstdlib>
#include <fstream>
#include <iostream>

#include "../db/pool/DBConnectionPool.h"
#include "../db/connection/DBConnection.hpp"
#include "../db/connection/DBExecutor.h"
#include "../db/table/Column.h"
#include "../utils/Utils.h"

#include "JsonGenerator.h"
#include "JsonToXml.h"
#include "RESTHandler.h"

namespace rest {

void RESTHandler::getResponse(const int responseCode, const std::string& body, restbed::Response& response) {
	// Set response headers
	response.add_header("Access-Control-Allow-Origin", "*");
	response.add_header("Access-Control-Allow-Methods", "POST, GET, OPTIONS, DELETE");
	response.add_header("Access-Control-Max-Age", "3600");
	response.add_header("Access-Control-Allow-Headers", "x-requested-with");
	// Set response code
	response.set_status_code(responseCode);
	// Set body
	response.set_body(body);

	response.add_header("Connection", "close");
}

void RESTHandler::databaseHandlerJson(const std::shared_ptr<restbed::Session> session) {
	databaseHandler(session, JSON);
}

void RESTHandler::databaseHandlerXml(const std::shared_ptr<restbed::Session> session) {
	databaseHandler(session, XML);
}

void RESTHandler::databaseHandler(const std::shared_ptr<restbed::Session> session, const formatType format) {
	if (session->is_open()) {
		try {
			std::vector<std::string> databases;
			db::DBExecutor::getDBs(databases);
			std::string body;
			rest::JsonGenerator::getJson("database", databases, body);

			restbed::Response response;
			if (format == JSON)
				getResponse(restbed::OK, body, response);
			else
				getResponse(restbed::OK, rest::JsonToXml::json2xml(body, "databases"), response);
			session->close(response);
		} catch (...) {
			std::cerr << "Internal server error" << std::endl;
			session->close(400, "Internal server error");
		}
	} else {
		std::cerr << "Internal server error" << std::endl;
		session->close(400, "Internal server error");
	}
}

void RESTHandler::tablesHandlerJson(const std::shared_ptr<restbed::Session> session) {
	tablesHandler(session, JSON);
}

void RESTHandler::tablesHandlerXml(const std::shared_ptr<restbed::Session> session) {
	tablesHandler(session, XML);
}

void RESTHandler::tablesHandler(const std::shared_ptr<restbed::Session> session, const formatType format) {
	if (session->is_open()) {
		try {
			std::string db;
			const std::shared_ptr<const restbed::Request> request = session->get_request();
			db = request->get_path_parameter("dbname", "/");

			std::vector<std::string> tables;
			db::DBExecutor::getTables(db, tables);
			std::string body;

			rest::JsonGenerator::getJson("table", tables, body);

			restbed::Response response;
			if (format == JSON)
				getResponse(restbed::OK, body, response);
			else
				getResponse(restbed::OK, rest::JsonToXml::json2xml(body, "tables"), response);
			session->close(response);
		} catch (...) {
			std::cerr << "Internal server error" << std::endl;
			session->close(400, "Internal server error");
		}
	} else {
		std::cerr << "Internal server error" << std::endl;
		session->close(400, "Internal server error");
	}
}

void RESTHandler::singleTableHandlerJson(const std::shared_ptr<restbed::Session> session) {
	singleTableHandler(session, JSON);
}

void RESTHandler::singleTableHandlerXml(const std::shared_ptr<restbed::Session> session) {
	singleTableHandler(session, XML);
}

void RESTHandler::singleTableHandler(const std::shared_ptr<restbed::Session> session, const formatType format) {
	if (session->is_open()) {
		try {
			const std::shared_ptr<const restbed::Request> request = session->get_request();
			const std::string dbName = request->get_path_parameter("dbname", "/");
			const std::string tableName = request->get_path_parameter("tablename", "/");

			std::shared_ptr<db::table::Column> table;
			db::DBExecutor::getTableInfo(dbName, tableName, table);

			std::string body;
			rest::JsonGenerator::getJson(tableName, table, body);

			restbed::Response response;
			if (format == JSON)
				getResponse(restbed::OK, body, response);
			else
				getResponse(restbed::OK, rest::JsonToXml::json2xml(body, "table"), response);
			session->close(response);
		} catch (...) {
			std::cerr << "Internal server error" << std::endl;
			session->close(400, "Internal server error");
		}
	} else {
		std::cerr << "Internal server error" << std::endl;
		session->close(400, "Internal server error");
	}
}

void RESTHandler::errorHandler(const int code, const std::exception& exc, const std::shared_ptr<restbed::Session> session) {
	throw restbed::SERVICE_UNAVAILABLE;
}

void RESTHandler::infoHandler(const std::shared_ptr<restbed::Session> session) {
	if (session->is_open()) {
		try {

			std::string body;
			body = "<h1>DBtoREST</h1>\n";
			body += "<p><h4>In this project I created a service that allows having access to all the information stored in a ";
			body += "database. More specifically, I have implemented a RESTful service that allows any user to know which ";
			body += "databases and tables exist in the configured database manager as well as show the information contained ";
			body += "in any table. However, in order to skip database manager tables, a filtering mechanism to not show them ";
			body += "has been implemented.</h4></p>\n";
			body += "Possible requests:\n";
			body += "<ul>\n";
			body += "<li>http://" + session->get_destination() + "/DBtoREST/alldbs</li>";
			body += "<li>http://" + session->get_destination() + "/DBtoREST/alltables/&lt;dbName&gt;</li>";
			body += "<li>http://" + session->get_destination() + "/DBtoREST/table/&lt;dbName&gt;/&lt;tableName&gt;</li>";
			body += "</ul>\n";

			restbed::Response response;
			response.add_header("Content-Type", "text/html");
			getResponse(restbed::OK, body, response);
			session->close(response);
		} catch (...) {
			std::cerr << "Internal server error" << std::endl;
			session->close(400, "Internal server error");
		}
	} else {
		std::cerr << "Internal server error" << std::endl;
		session->close(400, "Internal server error");
	}
}

void RESTHandler::swaggerJson(const std::shared_ptr<restbed::Session> session) {
	std::string body;
	std::string tmpStr;

	std::ifstream ifstr = std::ifstream(std::string("../swagger/swagger.json").c_str());
	while (std::getline(ifstr, tmpStr)) {
		body += tmpStr + "\n";
	}
	ifstr.close();
	session->set_headers({{"Access-Control-Allow-Methods", "POST, GET, OPTIONS, PUT"},{"Access-Control-Allow-Origin", "*"}});
	session->close(restbed::OK, body, { {"Connection", "close"} });
}

} /* namespace rest */

File: java/src/main/java/rest/RESTHandler.java
package main.java.rest;

import java.util.Vector;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import main.java.db.connection.DBExecutor;
import main.java.db.table.Column;
import main.java.utils.Utils;

/**
 * This classes defines the REST handlers
 * @author 	Ronald T. Fernandez
 * @version 1.0
 */
@Path("/")
public class RESTHandler {
	
	final static String SEP = System.getProperty("file.separator");
	final String CURR_DIR = this.getClass().getClassLoader().getResource("").getPath();
	
	/**
	 * Http allowed response codes
	 * OK: 	http code 200 (success)
	 * NOK: http code 404 (bad request)
	 */
	private enum HttpResponseCode {
		OK(200), 
		NOK(400);
		
		private int code;
		
		/**
		 * Class Constructor
		 * @param code	Http status value
		 */
		HttpResponseCode(int code) {
			this.code = code;
		}
		
		/**
		 * Get the associated status
		 * @return Http status
		 */
		public int getValue() {
			return this.code;
		}
	}
	
	private static Response getResponseWithHeaders(String content, HttpResponseCode status) {
		return Response.ok(content).
				status(status.getValue()).
				header("Access-Control-Allow-Origin","*").
				header("Access-Control-Allow-Methods", "POST, GET, OPTIONS, DELETE").
				header("Access-Control-Max-Age", "3600").
				header("Access-Control-Allow-Headers", "x-requested-with").build();
	}
	
	/**
	 * Publish the swagger settings file
	 * @return The swagger file content or, if an error occurs, the error message
	 */
	@GET
	@Path("/swagger.json")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getSwaggerJson() {
		try {
			final String fileName = CURR_DIR + SEP + ".." + SEP + ".." + SEP + "conf" + SEP + "swagger.json";
			return getResponseWithHeaders(Utils.readFile(fileName), HttpResponseCode.OK);
		} catch (Exception e) {
			return getResponseWithHeaders(e.getMessage(), HttpResponseCode.NOK);
		}
	}
	
	/**
	 * Define the handler when looking for the databases in the database manager in the JSON format
	 * @return The list of databases in JSON format or, if an error occurs, the error message
	 */
	@SuppressWarnings("static-method")
	@GET
	@Path("/alldbs")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getDBsJSON() {
		try {
			Vector<String> dbs = DBExecutor.getDBs();
			return getResponseWithHeaders(JSONGenerator.getJson("databases", dbs), HttpResponseCode.OK);
		} catch (Exception e) {
			return getResponseWithHeaders(e.getMessage(), HttpResponseCode.NOK);
		}
	}
	
	/**
	 * Define the handler when looking for the databases in the database manager in the XML format
	 * @return The list of databases in XML format or, if an error occurs, the error message
	 */
	@SuppressWarnings("static-method")
	@GET
	@Path("/alldbs")
	@Produces(MediaType.APPLICATION_XML)
	public Response getDBsXml() {
		try {
			Vector<String> dbs = DBExecutor.getDBs();
			return getResponseWithHeaders(JSONGenerator.getXml("databases", "database", dbs), HttpResponseCode.OK);
		} catch (Exception e) {
			return getResponseWithHeaders(e.getMessage(), HttpResponseCode.NOK);
		}
	}
	
	/**
	 * Define the handler when looking for the tables in a database in the JSON format
	 * @param dbName	Database name
	 * @return 			The list of tables in the database in JSON format or, if an error occurs, the error message
	 */
	@SuppressWarnings("static-method")
	@GET
	@Path("/alltables/{dbName}")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getTablesJson(@PathParam("dbName") String dbName) {
		try {
			Vector<String> tables = DBExecutor.getTables(dbName);
			return getResponseWithHeaders(JSONGenerator.getJson("tables", tables), HttpResponseCode.OK);
		} catch (Exception e) {
			return getResponseWithHeaders(e.getMessage(), HttpResponseCode.NOK);
		}
	}
	
	/**
	 * Define the handler when looking for the tables in a database in the XML format
	 * @param dbName	Database name
	 * @return 			The list of tables in the database in XML format or, if an error occurs, the error message
	 */
	@GET
	@Path("/alltables/{dbName}")
	@Produces(MediaType.APPLICATION_XML)
	public static Response getTablesXml(@PathParam("dbName") String dbName) {
		try {
			Vector<String> tables = DBExecutor.getTables(dbName);
			return getResponseWithHeaders(JSONGenerator.getXml("tables", "table", tables), HttpResponseCode.OK);
		} catch (Exception e) {
			return getResponseWithHeaders(e.getMessage(), HttpResponseCode.NOK);
		}
	}
	
	/**
	 * Define the handler when looking for a table content in the JSON format
	 * @param dbName	Database name
	 * @param tableName	Table name
	 * @return 			The content of the table in JSON format or, if an error occurs, the error message
	 */
	@GET
	@Path("/table/{dbName}/{tableName}")
	@Produces(MediaType.APPLICATION_JSON)
	public static Response getTableInfoJson(@PathParam("dbName") String dbName, @PathParam("tableName") String tableName) {
		try {
			Column table = DBExecutor.getTableInfo(dbName, tableName);
			return getResponseWithHeaders(JSONGenerator.getJson("table", table), HttpResponseCode.OK);
		} catch (Exception e) {
			return getResponseWithHeaders(e.getMessage(), HttpResponseCode.NOK);
		}
	}
	
	/**
	 * Define the handler when looking for a table content in the XML format
	 * @param dbName	Database name
	 * @param tableName	Table name
	 * @return The content of the table in XML format or, if an error occurs, the error message
	 */
	@GET
	@Path("/table/{dbName}/{tableName}")
	@Produces(MediaType.APPLICATION_XML)
	public static Response getTableInfoXml(@PathParam("dbName") String dbName, @PathParam("tableName") String tableName) {
		try {
			Column table = DBExecutor.getTableInfo(dbName, tableName);
			return getResponseWithHeaders(JSONGenerator.getXml("tables", "table", table), HttpResponseCode.OK);
		} catch (Exception e) {
			return getResponseWithHeaders(e.getMessage(), HttpResponseCode.NOK);
		}
	}
	
	/**
	 * Define the info handler: This is used as to show information about the project
	 * @return The information of the project in HTML format
	 */
	@GET
	@Path("/")
	@Produces(MediaType.TEXT_HTML)
	public static Response getInfo() {
		String host;
		try {
			host = Utils.getRESTHost();
		} catch (Exception e) {
			host = "127.0.0.1";
		}
		int port;
		try {
			port = Utils.getRESTPort();
		} catch (Exception e) {
			port = 8080;
		}
		
		String body = new String();
		body = "<h1>DBtoREST</h1>\n";
		body += "<p><h4>In this project I created a service that allows having access to all the information stored in a ";
		body += "database. More specifically, I have implemented a RESTful service that allows any user to know which ";
		body += "databases and tables exist in the configured database manager as well as show the information contained ";
		body += "in any table. However, in order to skip database manager tables, a filtering mechanism to not show them ";
		body += "has been implemented.</h4></p>\n";
		body += "Possible requests:\n";
		body += "<ul>\n";
		body += "<li>http://" + host + ":" + port + "/DBtoREST/alldbs</li>";
		body += "<li>http://" + host + ":" + port + "/DBtoREST/alltables/&lt;dbName&gt;</li>";
		body += "<li>http://" + host + ":" + port + "/DBtoREST/table/&lt;dbName&gt;/&lt;tableName&gt;</li>";
		body += "</ul>\n";
		
		return getResponseWithHeaders(body, HttpResponseCode.OK);
	}
	
}
